In [1]:
import numpy as np
import pandas as pd
from dask import dataframe as dd
import string
import re
import nltk
from nltk.stem.snowball import SnowballStemmer
from nltk.corpus import stopwords
from spacy.tokenizer import Tokenizer
from spacy.lang.en import English

nlp = English()
tokenizer = nlp.tokenizer
stemmer = SnowballStemmer("english")

pd.set_option('display.max_columns', 100)

# Government Spend Analysis

## Step 1: Examine column headers to identify columns of interest

In [2]:
test = pd.read_csv('FY2016_All_Contracts_Full_20210608_1.csv', nrows = 5)

test

contract_transaction_unique_key             contract_award_unique_key  \
0  8600_8600_0001_0_DU205NC16D02_0  CONT_AWD_0001_8600_DU205NC16D02_8600   
1      6800_6800_0001_5_EPR91502_0      CONT_AWD_0001_6800_EPR91502_6800   
2     6800_6800_0001_20_EPR51502_0      CONT_AWD_0001_6800_EPR51502_6800   
3      6800_6800_0001_0_EPS21602_0      CONT_AWD_0001_6800_EPS21602_6800   
4     6800_6800_0001_31_EPS51301_0      CONT_AWD_0001_6800_EPS51301_6800   

   award_id_piid  modification_number  transaction_number  \
0              1                    0                   0   
1              1                    5                   0   
2              1                   20                   0   
3              1                    0                   0   
4              1                   31                   0   

   parent_award_agency_id                      parent_award_agency_name  \
0                    8600  HOUSING AND URBAN DEVELOPMENT, DEPARTMENT OF   
1                    6800               ENVIRONMENTAL PROTECTION AGENCY   
2                    6800               ENVIRONMENTAL PROTECTION AGENCY   
3                    6800               ENVIRONMENTAL PROTECTION AGENCY   
4                    6800               ENVIRONMENTAL PROTECTION AGENCY   

  parent_award_id_piid  parent_award_modification_number  \
0         DU205NC16D02                                 0   
1             EPR91502                                 0   
2             EPR51502                                 0   
3             EPS21602                                 0   
4             EPS51301                                 0   

   federal_action_obligation  total_dollars_obligated  \
0                  799750.00                546621.23   
1                       0.00                      NaN   
2                  750000.00               3042661.55   
3                 1149194.00                      NaN   
4                   99339.58                      NaN   

   base_and_exercised_options_value  current_total_value_of_award  \
0                         799750.00                     546621.23   
1                              0.00                           NaN   
2                         750000.00                   16600000.00   
3                        1389501.48                           NaN   
4                          99339.58                           NaN   

   base_and_all_options_value  potential_total_value_of_award  \
0                   799750.00                       546621.23   
1                        0.00                             NaN   
2                   750000.00                     16600000.00   
3                  1389501.48                             NaN   
4                    99339.58                             NaN   

   disaster_emergency_fund_codes_for_overall_award  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   

   outlayed_amount_funded_by_COVID-19_supplementals_for_overall_aw  \
0                                                NaN                 
1                                                NaN                 
2                                                NaN                 
3                                                NaN                 
4                                                NaN                 

   obligated_amount_funded_by_COVID-19_supplementals_for_overall_a  \
0                                                NaN                 
1                                                NaN                 
2                                                NaN                 
3                                                NaN                 
4                                                NaN                 

  action_date  action_date_fiscal_year period_of_pe

### Columns of interest:

* contract_transaction_unique_key
* contract_award_unique_key
* awarding_agency_code
* awarding_sub_agency_code
* awarding_office_code
* funding_agency_code
* funding_sub_agency_code
* funding_office_code
* recipient_duns
* recipient_name
* recipient_doing_business_as_name
* recipient_parent_duns


## Step 2: Load data

Using dask with only columns of interest included.

In [3]:
columns = ['contract_transaction_unique_key', 'contract_award_unique_key', 'awarding_agency_code', 'awarding_sub_agency_code',\
           'awarding_office_code', 'funding_agency_code', 'funding_sub_agency_code', 'funding_office_code', 'recipient_duns', \
           'recipient_name', 'recipient_doing_business_as_name', 'recipient_parent_duns']

data = dd.read_csv('*.csv', usecols=columns, dtype={'awarding_sub_agency_code': 'object', 'funding_sub_agency_code': 'object', \
                                                    'recipient_doing_business_as_name': 'object', 'recipient_parent_duns': 'object',\
                                                    'funding_agency_code': 'object', 'awarding_office_code': 'object', \
                                                    'funding_office_code': 'object', 'recipient_duns': 'object'})

data.head()

contract_transaction_unique_key             contract_award_unique_key  \
0  8600_8600_0001_0_DU205NC16D02_0  CONT_AWD_0001_8600_DU205NC16D02_8600   
1      6800_6800_0001_5_EPR91502_0      CONT_AWD_0001_6800_EPR91502_6800   
2     6800_6800_0001_20_EPR51502_0      CONT_AWD_0001_6800_EPR51502_6800   
3      6800_6800_0001_0_EPS21602_0      CONT_AWD_0001_6800_EPS21602_6800   
4     6800_6800_0001_31_EPS51301_0      CONT_AWD_0001_6800_EPS51301_6800   

   awarding_agency_code awarding_sub_agency_code awarding_office_code  \
0                    86                     8600               86545A   
1                    68                     6800               68HE09   
2                    68                     6800               68HE05   
3                    68                     6800               68HE02   
4                    68                     6800               68HE05   

  funding_agency_code funding_sub_agency_code funding_office_code  \
0                 086                    8600              8630AX   
1                 068                    6800              68Y000   
2                 068                    6800              68U000   
3                 068                    6800              68R000   
4                 068                    6800              68U000   

  recipient_duns                                     recipient_name  \
0      831854596    MORTGAGE ASSISTANCE AND ACQUISITIONS GROUP, LLC   
1      112621987                                       INNOVATE INC   
2      079493355  RYBA MARINE CONSTRUCTION CO/TERRA CONTRACTING ...   
3      080265258                                       CAT-ENSOFTEK   
4      178599221                                TETRA TECH EM, INC.   

  recipient_doing_business_as_name recipient_parent_duns  
0                              NaN             831854596  
1                              NaN             112621987  
2                              NaN             079493355  
3                              NaN             080265258  
4                              NaN             045224250

### Step 2a: Save as parquet format

(So it's easier to reload if needed)

In [4]:
data.to_parquet('government_spend_dataset.parquet', engine='pyarrow')

(None,)

## Step 3: Get number of initial records

In [5]:
a = data.index.size.compute()

print("There are ", a, " initial records in this dataset.")

There are  4820290  initial records in this dataset.


## Step 4: Create list of unique suppliers (exact)

(and get number of unique suppliers)

In [6]:
suppliers_unique = data.recipient_name.unique().compute().reset_index()

print("There are ", len(suppliers_unique), " unique suppliers (exact match only).")

There are  153462  unique suppliers (exact match only).


In [7]:
suppliers_unique.head()

index                                     recipient_name
0      0    MORTGAGE ASSISTANCE AND ACQUISITIONS GROUP, LLC
1      1                                       INNOVATE INC
2      2  RYBA MARINE CONSTRUCTION CO/TERRA CONTRACTING ...
3      3                                       CAT-ENSOFTEK
4      4                                TETRA TECH EM, INC.

### Step 4a: Does recipient_doing_business_as_name have any non-null values?

Answer: yes, hold off on this until next step

In [8]:
data.recipient_doing_business_as_name.dropna().compute()

23                MIT LINCOLN LABORATORY
29                                 U T C
39        SOFTWARE ENGINEERING INSTITUTE
52                    PACIFIC SCIENTIFIC
86                    PACIFIC SCIENTIFIC
                      ...               
18863    TOUCHSTONE NEURORECOVERY CENTER
18864                     PROVIDE A RIDE
18875             PETNET PHARMACEUTICALS
18882        MUSC DEPARTMENT OF MEDICINE
18895                 BIMBO BAKERIES USA
Name: recipient_doing_business_as_name, Length: 94562, dtype: object

## Step 5: Normalize text

Using:

 - list of punctuation from string package
 - custom regex to remove spaces between single character words
 - spacy English tokenizer
 - Snowball stemmer from NLTK
 - English stopwords from NLTK

Returning normalized tokens as list in additional column

In [199]:
stop = ['the', 'and', 'inc', 'llc', 'co', 'company', 'corporation', 'group', 'lda', 'services']
punct = string.punctuation

def remove_punct(text):
    text_str = str(text)
    for c in text_str:
        if c in punct:
            text_str = text_str.replace(c, '')
    return text_str

def remove_whitespace(text_str):
    text_str = re.sub(r'(?<=\s[A-Z0-9])\s*[ ]\s*(?=[A-Z0-9]\s)', '', text_str)
    text_str = re.sub(r'(?<=^[A-Z0-9])\s*[ ]\s*(?=[A-Z0-9])', '', text_str)
    text_str = re.sub(r'(?<=[A-Z0-9])\s*[ ]\s*(?=[A-Z0-9]$)', '', text_str)
    return text_str

def normalize_text(text_str):
    wordlist = []
    for w in tokenizer(text_str):
        if str(w).lower() not in stop and not w.is_punct and not w.is_space:
            wordlist.append(stemmer.stem(str(w).lower()))
    return wordlist

suppliers_unique['recipient_name_norm'] = suppliers_unique.recipient_name.\
apply(remove_punct).apply(remove_whitespace).apply(normalize_text)

suppliers_unique

index                                     recipient_name  \
0            0    MORTGAGE ASSISTANCE AND ACQUISITIONS GROUP, LLC   
1            1                                       INNOVATE INC   
2            2  RYBA MARINE CONSTRUCTION CO/TERRA CONTRACTING ...   
3            3                                       CAT-ENSOFTEK   
4            4                                TETRA TECH EM, INC.   
...        ...                                                ...   
153457  153457           AZORMED - GESTAO AMBIENTAL ACOREANA, LDA   
153458  153458                                      JEAN ROB & CO   
153459  153459                               WOODS SERVICES, INC.   
153460  153460                               FLOWERS & ASSOCIATES   
153461  153461                                       GARY BLUEMKE   

                                      recipient_name_norm  \
0                             [mortgag, assist, acquisit]   
1                                                 [innov]   
2       [ryba, marin, construct, coterra, contract, a,...   
3                                           [catensoftek]   
4                                       [tetra, tech, em]   
...                                                   ...   
153457                 [azorm, gestao, ambient, acoreana]   
153458                                        [jean, rob]   
153459                                             [wood]   
153460                                   [flower, associ]   
153461                                     [gari, bluemk]   

                                       recipient_name_str  
0                                 mortgag,assist,acquisit  
1                                                   innov  
2       ryba,marin,construct,coterra,contract,a,joint,...  
3                                             catensoftek  
4                                           tetra,tech,em  
...                                                   ...  
153457                      azorm,gestao,ambient,acoreana  
153458                                           jean,rob  
153459                                               wood  
153460                                      flower,associ  
153461                                        gari,bluemk  

[153462 rows x 4 columns]

In [200]:
suppliers_unique['recipient_name_str'] = suppliers_unique.recipient_name_norm.agg(lambda x: ','.join(map(str, x)))

suppliers_unique.recipient_name_str.value_counts()[suppliers_unique.recipient_name_str.value_counts() > 1].\
sort_values(ascending = False).to_frame(name = 'Additional matches')

Additional matches
advanc,prosthet,orthot                   9
analyt                                   8
ur                                       6
apprais                                  6
advanc,technolog                         6
...                                    ...
sweat                                    2
air,condit,mainten                       2
dtech,lab                                2
qwaltec,incorpor                         2
pgba                                     2

[6488 rows x 1 columns]

There were 6,488 additional matches discovered through this simple text normalization.

In [201]:
suppliers_unique[suppliers_unique.recipient_name_str == "analyt"]

index               recipient_name recipient_name_norm  \
15135    15135       ANALYTIC SERVICES INC.            [analyt]   
19186    19186    ANALYTICAL SERVICES, INC.            [analyt]   
25627    25627        ANALYTIC SERVICES INC            [analyt]   
52629    52629       ANALYTIC SERVICES, INC            [analyt]   
67509    67509              ANALYTICS, INC.            [analyt]   
83935    83935  ANALYTICAL SERVICES COMPANY            [analyt]   
97088    97088     ANALYTICAL SERVICES, INC            [analyt]   
100001  100001        ANALYTICS CORPORATION            [analyt]   

       recipient_name_str  
15135              analyt  
19186              analyt  
25627              analyt  
52629              analyt  
67509              analyt  
83935              analyt  
97088              analyt  
100001             analyt

In [202]:
suppliers_unique.recipient_name_str.unique().shape

suppliers_unique_list = suppliers_unique.recipient_name_str.unique()

In [203]:
len(suppliers_unique_list)

146202

With normalized text, **146,202** unique suppliers are found. These new normalized names can be joined to the original df. The right index is a unique id per supplier.

In [204]:
data_norm = data.merge(suppliers_unique, on = 'recipient_name').compute()

data_norm

contract_transaction_unique_key  \
0                  8600_8600_0001_0_DU205NC16D02_0   
1                      6800_6800_0001_5_EPR91502_0   
2                      6800_6800_0002_3_EPR91502_0   
3                     6800_6800_0001_20_EPR51502_0   
4                      6800_6800_0001_0_EPS21602_0   
...                                            ...   
18914  7008_-NONE-_HSCG3216PSMA006_P00001_-NONE-_0   
18915    9700_-NONE-_SPE4A716D5005_0_-NONE-_-NONE-   
18916     3600_-NONE-_VA25815P0001_P00002_-NONE-_0   
18917    9700_-NONE-_SPE4A716D0005_0_-NONE-_-NONE-   
18918        1605_-NONE-_DOLJ089626807_31_-NONE-_0   

                         contract_award_unique_key  awarding_agency_code  \
0             CONT_AWD_0001_8600_DU205NC16D02_8600                    86   
1                 CONT_AWD_0001_6800_EPR91502_6800                    68   
2                 CONT_AWD_0002_6800_EPR91502_6800                    68   
3                 CONT_AWD_0001_6800_EPR51502_6800                    68   
4                 CONT_AWD_0001_6800_EPS21602_6800                    68   
...                                            ...                   ...   
18914  CONT_AWD_HSCG3216PSMA006_7008_-NONE-_-NONE-                    70   
18915                  CONT_IDV_SPE4A716D5005_9700                    97   
18916     CONT_AWD_VA25815P0001_3600_-NONE-_-NONE-                    36   
18917                  CONT_IDV_SPE4A716D0005_9700                    97   
18918    CONT_AWD_DOLJ089626807_1605_-NONE-_-NONE-                  1601   

      awarding_sub_agency_code awarding_office_code funding_agency_code  \
0                         8600               86545A                 086   
1                         6800               68HE09                 068   
2                         6800               68HE09                 068   
3                         6800               68HE05                 068   
4                         6800               68HE02                 068   
...                        ...                  ...                 ...   
18914                     7008                00032                 070   
18915                     97AS               SPE4A7                 097   
18916                     3600                00258                 036   
18917                     97AS               SPE4A7                 097   
18918                     1630                ETATP                1601   

      funding_sub_agency_code funding_office_code recipient_duns  \
0                        8600              8630AX      831854596   
1                        6800              68Y000      112621987   
2                        6800              68Y000      112621987   
3                        6800              68U000      079493355   
4                        6800              68R000      080265258   
...                       ...                 ...            ...   
18914                    7008               00032      785365693   
18915                    97AS              SPE4A7      089107726   
18916                    3600               00678      968237433   
18917                    97AS              SPE4A7      121360627   
18918                    1631               ETATP      145057522   

                                          recipient_name  \
0        MORTGAGE ASSISTANCE AND ACQUISITIONS GROUP, LLC   
1                                           INNOVATE INC   
2                                           INNOVATE INC   
3      RYBA MARINE CONSTRUCTION CO/TERRA CONTRACTING ...   
4                                           CAT-ENSOFTEK   
...                                                  ...   
18914                                   FLO-TECH, L.L.C.   
18915                  PRECI-MANUFACTURING, INCORPORATED   
18916                                       GARY BLUEMKE   
18917                           JANEL'S INDUSTRIES, INC.   
18918           UBC NATIONAL JOB CORPS TRAINING FUND INC   

      recipient_do

### Step 5a: Normalizing and using recipient_doing_business_as_name column

Names also appear for some rows in recipient_doing_business_as_name. For rows where this appears, this can also be normalized and used in place of normed values in recipient_name column in cases where it may lead to a larger number of matches.

In [205]:
suppliers_unique_rev = data_norm.recipient_doing_business_as_name.dropna().drop_duplicates().reset_index()

suppliers_unique_rev

index       recipient_doing_business_as_name
0       109                 MIT LINCOLN LABORATORY
1       136                                 BOEING
2       158                                  U T C
3       224         SOFTWARE ENGINEERING INSTITUTE
4       238        INTERNATIONAL BUSINESS MACHINES
...     ...                                    ...
8893  18571        GOLDEN LIVINGCENTER - BRIARWOOD
8894  18756                   AVALON PLACE WHARTON
8895  18758       BRIARWOOD NURSING&REHABILITATION
8896  18849                FAIRFIELD INN -MARRIOTT
8897  18875  HOPE NETWORK TRAUMATIC BRAIN INJURIES

[8898 rows x 2 columns]

In [214]:
suppliers_unique_rev['recipient_doing_business_norm'] = suppliers_unique_rev.recipient_doing_business_as_name.apply(remove_punct).\
apply(remove_whitespace).\
apply(normalize_text)

suppliers_unique_rev['recipient_doing_business_str'] = suppliers_unique_rev.recipient_doing_business_norm.agg(lambda x: ','.join(map(str, x)))

In [213]:
data_norm_rev = data_norm.merge(suppliers_unique_rev, how = 'outer', \
                                on = ['recipient_doing_business_as_name'])

In [208]:
suppliers_rev = data_norm_rev[['recipient_name', 'recipient_name_str', 'recipient_doing_business_str']].drop_duplicates().reset_index()

suppliers_business_as_mult = suppliers_rev.recipient_doing_business_str.dropna()\
[suppliers_rev.recipient_doing_business_str.map(suppliers_rev.\
                                                recipient_doing_business_str.value_counts()) > 1].\
drop_duplicates().\
tolist()


In [211]:
suppliers_rev['recipient_str_rev'] = np.where(((suppliers_rev['recipient_doing_business_str'].\
                                               isin(suppliers_business_as_mult))|\
                                               (suppliers_rev['recipient_doing_business_str'].isin(suppliers_unique))), \
suppliers_rev['recipient_doing_business_str'],\
suppliers_rev['recipient_name_str'])


In [240]:
suppliers_rev['recipient_str_rev'].drop_duplicates().shape[0]

146182

This technique offered a slight additional improvement, with the result that **146,182** unique suppliers are found.

In [216]:
suppliers_rev = suppliers_rev[['recipient_str_rev', 'recipient_name']].reset_index()

data_final = data.merge(suppliers_rev, how = 'outer', on = 'recipient_name').compute()

data_final

contract_transaction_unique_key  \
0     1605_1605_DOLOPS15U00186_4_DOLQ119431629_0   
1     1605_1605_DOLOPS15U00201_2_DOLQ119431629_0   
2     1605_1605_DOLOPS15U00186_3_DOLQ119431629_0   
3     1605_1605_DOLOPS16U00149_0_DOLQ119431629_0   
4      1605_1605_DOLU119K32631_2_DOLQ119431629_0   
...                                          ...   
9055    12C2_-NONE-_AG03R6P1568016905_0_-NONE-_0   
9056   9700_9700_FA462516FG108_0_FA462515A0004_0   
9057      9700_4732_FA468616FG018_0_GS03F055CA_0   
9058        3600_-NONE-_VA118A15P0388_0_-NONE-_0   
9059    3600_-NONE-_VA26114C0120_P00002_-NONE-_0   

                            contract_award_unique_key  awarding_agency_code  \
0     CONT_AWD_DOLOPS15U00186_1605_DOLQ119431629_1605                  1601   
1     CONT_AWD_DOLOPS15U00201_1605_DOLQ119431629_1605                  1601   
2     CONT_AWD_DOLOPS15U00186_1605_DOLQ119431629_1605                  1601   
3     CONT_AWD_DOLOPS16U00149_1605_DOLQ119431629_1605                  1601   
4      CONT_AWD_DOLU119K32631_1605_DOLQ119431629_1605                  1601   
...                                               ...                   ...   
9055    CONT_AWD_AG03R6P1568016905_12C2_-NONE-_-NONE-                    12   
9056   CONT_AWD_FA462516FG108_9700_FA462515A0004_9700                    97   
9057      CONT_AWD_FA468616FG018_9700_GS03F055CA_4732                    97   
9058        CONT_AWD_VA118A15P0388_3600_-NONE-_-NONE-                    36   
9059         CONT_AWD_VA26114C0120_3600_-NONE-_-NONE-                    36   

     awarding_sub_agency_code awarding_office_code funding_agency_code  \
0                        1605               1605DC                1601   
1                        1605               1605DC                1601   
2                        1605               1605DC                1601   
3                        1605               1605DC                1601   
4                        1605               1605DC                1601   
...                       ...                  ...                 ...   
9055                     12C2                 03R6                 012   
9056                     5700               FA4625                 097   
9057                     5700               FA4686                 097   
9058                     3600                0010A                 036   
9059                     3600                00261                 036   

     funding_sub_agency_code funding_office_code recipient_duns  \
0                       1667              1667DC      083634829   
1                       1667              1667DC      083634829   
2                       1667              1667DC      083634829   
3                       1667              1667DC      083634829   
4                       1605              160501      083634829   
...                      ...                 ...            ...   
9055                    12C2                0351      007978216   
9056                    5700              F4E32H      031164965   
9057                    5700              F1H2EE      869288274   
9058                    3600               0010A      610226243   
9059                    3600               00261      077668325   

                                      recipient_name  \
0                                       CONCEPTS INC   
1                                       CONCEPTS INC   
2                                       CONCEPTS INC   
3                                       CONCEPTS INC   
4                                       CONCEPTS INC   
...                                              ...   
9055                               HAHN SUPPLY, INC.   
9056  SWEENEY-PHILLIPS & HOLDREN FUNERAL SERVICE INC   
9057                   CHAMPLIN TIRE RECYCLING, INC.   
9058                           SERENGETI SYSTEMS INC   
9059                        MOLOKAI GENERAL HOSPITAL   

     recipient_doing_business_as_name recipient_parent_duns   index  \
0    

## Step 6: Evaluating accuracy

I assume that the recipient_parent_duns column provides the true unique suppliers in this dataset and evaluate accuracy of the text normalization approach using the recipient_parent_duns column to provide a few different accuracy measures: 

- **Precision**
- **Recall**
- **F1 Score**

In [217]:
data_final['category'] = data_final['index']

accuracy = data_final[['recipient_parent_duns', 'category']].drop_duplicates().reset_index()

index recipient_parent_duns  category
0           0             083634829     13994
1          33             118341234      1399
2         226             625405191     19669
3         246             049591852     19773
4         247             370407639     11550
...       ...                   ...       ...
161853   9055             007978216    147337
161854   9056             031164965    148354
161855   9057             869288274    148366
161856   9058             610226243    148683
161857   9059             077668325    148781

[161858 rows x 3 columns]

### Precision

In [237]:
precision = accuracy.category[accuracy.category.map(accuracy.category.value_counts()) == 1].shape[0]/\
(accuracy.category[accuracy.category.map(accuracy.category.value_counts()) == 1].shape[0] +\
accuracy.category[accuracy.category.map(accuracy.category.value_counts()) > 1].shape[0])

print('The precision of this approach is', str(round(precision, 2)), '.')

The precision of this approach is 0.97 .


### Recall

In [238]:
sensitivity = accuracy.category[accuracy.category.map(accuracy.category.value_counts()) == 1].shape[0]/\
(accuracy.category[accuracy.category.map(accuracy.category.value_counts()) == 1].shape[0] +\
accuracy.recipient_parent_duns[accuracy.recipient_parent_duns.map(accuracy.recipient_parent_duns.value_counts()) > 1].shape[0])

print('The recall of this approach is', str(round(sensitivity, 2)), '.')

The recall of this approach is 0.79 .


### F1 Score

In [239]:
f1 = ((accuracy.category[accuracy.category.map(accuracy.category.value_counts()) == 1].shape[0])*2)/\
(((accuracy.category[accuracy.category.map(accuracy.category.value_counts()) == 1].shape[0])*2) +\
 (accuracy.category[accuracy.category.map(accuracy.category.value_counts()) > 1].shape[0]) +\
 (accuracy.recipient_parent_duns[accuracy.recipient_parent_duns.map(accuracy.recipient_parent_duns.value_counts()) > 1].shape[0]))

print('The F1 Score from this approach is', str(round(f1, 2)), '.')
 

The F1 Score from this approach is 0.87 .


### Step 6a: Alternative strategies

Instead of going the route of text normalization, conversion of sentence embeddings could be implemented to allow more flexibility and fine-tuning in the fuzzy matching procedure. With every word transformed into a numeric value, the resulting list per row could be compared to other rows in terms of a threshold cosine similarity or Levenshtein distance and fuzzy matching could be performed that way. The other benefit to this tactic would the ability for the analyst to control the matching threshold value. I initially thought about pursuing this strategy instead of or in addition to the normalization pipeline I ultimately used, but I did not have enough time to think about how to build a function that would do this pairwise calculation quickly between every possible pair of supplier names in this dataset, perhaps through k-means or other naive clustering analysis (resulting category labels could be appended to original df). It would be interesting to see how the accuracy of this strategy would compare to the strategy I ultimately used (the text normalization strategy had high precision but lower recall, while a strategy such as this may have higher recall but lower precision).

## Step 7: Newly answerable questions